### PR TITLE
feat(lookup/merge_variables): Add all hosts mode to collect configuration across multiple hosts

### PR DESCRIPTION
**Pull request by alpex8**
##### SUMMARY

This modifies the lookup plugin merge_variables. It's adding an all_hosts mode which allows to merge variables which have their pieces distributed across multiple hosts.  
I would like to give an example where this is useful. Lets start with:

* A generic database host without any further database/user configuration deployed via a role that is also able to maintain databases, users etc.
* Another host running a web service which is using the database service running on the remote host

How could you provide the configuration for the service to the database? The current merge_variables lookup plugin does not help here, because it's limited to the variables of the same host.

###### Configuration options

* You could add the configuration directly to the database host.
  * Either as host_vars or via a custom group
  * You may have the information on database name, user, password redundantly
* You could use hostvars['<service_host>'] from the database host to access its variables
* You could kind of create your own semantic and create a group per service and collect configuration elements from all those groups like:
  * ```yaml
    # db1-host.yml

    _default_databases:
      - name: db1
        owner: db1owner
    pg_databases:
      "{{ 
        _default_databases + 
        _service_a_databases | default([]) +
        _service_b_databases | default([]) +
        .... 
      }}"

    # service-a-host.yml

    _service_a_databases:
      - name: srv_a_db
        owner: srv_a_owner
    ```

There may be other strategies to handle this, but this is what I did so far in my projects. In the end at scale they all turned out to be cumbersome and difficult to maintain, because I felt forced to hop around too much files.

###### Cross Host Configuration Discovery

With the extended lookup I can add the configuration for the database, user, etc. where I need it - which where the whole service configuration is defined. The database role will use the lookup to pick up the relevant pieces of that configuration and use it for the configuration of the database instance.

```yaml
  # db1-host.yml

  - name: Discover Database Configuration
    set_fact:
      pg_databases: "{{ (lookup('community.general.merge_variables', 'database_config__', pattern_type='prefix', all_hosts=True)).databases }}"

  # service-a-host
  database_config__service_a_databases:
    target_inventory_hostname: db1-host
    databases:
      - name: srv_a_db
        owner: srv_a_owner

  # service-a-host
  database_config__service_b_databases:
    target_inventory_hostname: db1-host
    databases:
      - name: srv_b_db
        owner: srv_b_owner
```

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

plugins/lookup/merge_variables

##### ADDITIONAL INFORMATION

To the existing merge_variables lookup I introduced two additional parameters:

* all_hosts - To run the lookup plugin in all hosts mode
* target_host_property 
  * This property is to indicate how the optional key is named that controls which hosts are allowed to pick up configuration pieces.
  * It's default value is 'target_inventory_hostname'.
  * If configuration pieces does not have this property set, those pieces are only allowed to be picked up by the same host, which is basically the behavior when all_hosts mode is turned off. But to avoid having two lookup invocations (same host with all_hosts = False, other hosts with all_hosts = True) I added this for convenience.

For the all_hosts mode I tried to reuse as much as possible of the existing code. The added unit tests are using the all_hosts mode, but in that context I didn't really found out how to test the added feature. Thus the idea of that unit tests is more to show that the extension didn't break something.

### CODE REVIEW

**Comment by alpex8** [comment]
I had to remove myself as an author because that made some sanity check fail. I guess that's intended?

**Comment by felixfontein** [comment]
@alpex8 if you add yourself as an author, you also have to add yourself as a maintainer. See `CONTRIBUTING.md` and/or `.github/BOTMETA.yml`.

File: plugins/lookup/merge_variables.py [hunk 1]
```diff
@@ -61,6 +61,27 @@
         ini:
           - section: merge_variables_lookup
             key: override
+      all_hosts:
+        description:
+          - Search for variables accross all hosts. This mode allows to collect configuration pieces
+             accross different hosts (e.g. a service on a host with its database on another host).
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:67]:
```suggestion
             accross different hosts (for example a service on a host with its database on another host).
```

File: plugins/lookup/merge_variables.py [hunk 2]
```diff
@@ -61,6 +61,27 @@
...
+          - Each of the matching variables may container a property which specified for which host(s)
+             it is intended (see parameter 'target_host_property').
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:69]:
```suggestion
             it is intended (see parameter O(target_host_property)).
```

File: plugins/lookup/merge_variables.py [hunk 3]
```diff
@@ -61,6 +61,27 @@
...
+          - If that property isn't included in the matching variable, it only applies for the same host.
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:70]:
```suggestion
          - If that property is not included in the matching variable, it only applies for the same host.
```

File: plugins/lookup/merge_variables.py [hunk 4]
```diff
@@ -61,6 +61,27 @@
...
+        type: bool
+        default: false
+        choices:
+          - True
+          - False
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:75]: Booleans have two possible values, there is no need to provide `choices`. You need to provide `version_added` though:
```suggestion
        version_added: 8.4.0
```

File: plugins/lookup/merge_variables.py [hunk 5]
```diff
@@ -61,6 +61,27 @@
...
+      target_host_property:
+        description:
+          - When in all_hosts mode specifies the name of the property which contains the
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:78]:
```suggestion
          - When O(all_hosts=true) specifies the name of the property which contains the
```

File: plugins/lookup/merge_variables.py [hunk 6]
```diff
@@ -61,6 +61,27 @@
...
+             inventory hostname property of the host(s) for which the variable content is intended.
+          - The property can either be given as a string to address one host or a list of strings to
+             address multiple hosts.
+          - Each value must match the inventory_hostname property of the corresponding destination hosts.
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:82]:
```suggestion
          - Each value must match the C(inventory_hostname) property of the corresponding destination hosts.
```

File: plugins/lookup/merge_variables.py [hunk 7]
```diff
@@ -61,6 +61,27 @@
...
+        type: str
+        default: 'target_inventory_hostname'
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:84]:
```suggestion
        default: 'target_inventory_hostname'
        version_added: 8.4.0
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Some first comments on the docs.

**Comment by alpex8** [comment]
Thanks for the comments! I made the suggested changes.

**Comment by m-a-r-k-e** [comment]
Hi @alpex8,

Thanks for your contribution to this module, much appreciated :)

The idea for this pull request is clear and we see the benefit behind the idea (merging out-of-scope variables).

Currently, the `all_hosts` property makes it specific so that only everything can be merged (or with `all_hosts=False` the current behavior of merging your own variables). However, we think the plugin would be more flexible if groups can be specified instead of only all hosts.

Here a few examples:

```yml
# Default behavior (no extended scope functionality)
pg_databases: "{{ lookup('community.general.merge_variables', 'db_config__', pattern_type='prefix', groups=None) }}"

# Your suggested behavior, for merging variables for all hosts
pg_databases: "{{ lookup('community.general.merge_variables', 'db_config__', pattern_type='prefix', groups=['all']) }}"

# Custom behavior to only merge variables for specific groups
pg_databases: "{{ lookup('community.general.merge_variables', 'db_config__', pattern_type='prefix', groups=['test_group1', 'test_group2']) }}"
```

We think the addition of the `target_inventory_hostname` can be less complex and without having impact on the data structure that will be merged (currently you require a `dict` if in `all_hosts` mode).

Based on your described and desired behavior, the following would be possible without the need for the `target_inventory_hostname` option.

Option 1: Using of a unique prefix (with `groups=['all']`), e.g.

```yml
# roles/db-server-role/defaults/main.yml
# Use a default to allow merging of everything, this can be made more specific if needed.
pg_database_prefix: 'db_config__'

# roles/db-server-role/vars/main.yml
# Use the specified prefix to filter
pg_databases: "{{ lookup('community.general.merge_variables', pg_database_prefix, pattern_type='prefix', groups=['all']) }}"

# host_vars/db-host-a.yml
pg_database_prefix: 'db_config_a__'

# host_vars/db-host-b.yml
pg_database_prefix: 'db_config_b__'

# host_vars/service-1-host.yml
db_config_a__databases:
  - name: srv_1_db
    owner: srv_1_owner

# host_vars/service-2-host.yml
db_config_a__databases:
  - name: srv_2_db
    owner: srv_2_owner

# host_vars/service-3-host.yml
db_config_b__databases:
  - name: srv_3_db
    owner: srv_3_owner
```

Option 2: Use specific groups as scope.

```yml
# roles/db-server-role/defaults/main.yml
# Use the 'all' group as default, this can be overridden
pg_database_config_groups:
  - 'all'

# roles/db-server-role/vars/main.yml
# Use the specified prefix to filter
pg_databases: "{{ lookup('community.general.merge_variables', 'db_config__', pattern_type='prefix', groups=pg_database_config_groups) }}"

# host_vars/db-host-a.yml
pg_database_config_groups: 'db_a_hosts'

# host_vars/db-host-b.yml
pg_database_config_groups: 'db_b_hosts'

# host_vars/service-1-host.yml
db_config__databases:
  - name: srv_1_db
    owner: srv_1_owner

# host_vars/service-2-host.yml
db_config__databases:
  - name: srv_2_db
    owner: srv_2_owner

# host_vars/service-3-host.yml
db_config__databases:
  - name: srv_3_db
    owner: srv_3_owner

# inventory/ansible-inventory.ini
[db_a_hosts]
service-1-host
service-2-host

[db_b_hosts]
service-3-host
```

Based on these suggestions we think the implementation is more generic and has less limitations.

**Comment by alpex8** [comment]
> Hi @alpex8,
> 
> Thanks for your contribution to this module, much appreciated :)
> 
> The idea for this pull request is clear and we see the benefit behind the idea (merging out-of-scope variables).
> 
> Currently, the `all_hosts` property makes it specific so that only everything can be merged (or with `all_hosts=False` the current behavior of merging your own variables). However, we think the plugin would be more flexible if groups can be specified instead of only all hosts.
> 
> [...]
>
> Based on these suggestions we think the implementation is more generic and has less limitations.

Thanks for your suggestions. I like the idea and got rid of the additional property in the variables. I adjusted the feature according to your review.

File: plugins/lookup/merge_variables.py [hunk 8]
```diff
@@ -61,6 +62,13 @@
         ini:
           - section: merge_variables_lookup
             key: override
+      groups:
+        description:
+          - Search for variables accross hosts that belong to the given groups. This allows to collect configuration pieces
+             accross different hosts (for example a service on a host with its database on another host).
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:68]:
```suggestion
            accross different hosts (for example a service on a host with its database on another host).
```

File: plugins/lookup/merge_variables.py [hunk 9]
```diff
@@ -61,6 +62,13 @@
...
+        type: list
+        elements: str
+        version_added: 8.4.0
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:71]:
```suggestion
        version_added: 8.5.0
```
8.4.0 is already released.

File: plugins/lookup/merge_variables.py [hunk 10]
```diff
@@ -131,22 +139,52 @@ def _verify_and_get_type(variable):
...
                 raise AnsibleError("Non-string type '{0}' passed, only 'str' types are allowed!".format(type(term)))
 
-            ret.append(self._merge_vars(term, initial_value, variables))
+            if not self._groups:  # consider only own variables
+                ret.append(self._merge_vars(term, initial_value, variables))
+            else:  # consider variables of hosts in given groups
+                cross_host_merge_result = None
+                cross_host_var_type = ""
+                for host in variables["hostvars"]:
+                    if self._is_host_in_allowed_groups(variables["hostvars"][host]['group_names']):
```
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:160]: The other surrounding lines use double quotes:
```suggestion
                    if self._is_host_in_allowed_groups(variables["hostvars"][host]["group_names"]):
```

**Comment by alpex8** [comment]
Thanks for your hints @felixfontein. I took them over into the branch.

File: plugins/lookup/merge_variables.py [hunk 11]
```diff
@@ -131,22 +139,52 @@ def _verify_and_get_type(variable):
...
+                                cross_host_merge_result = host_result
+                                cross_host_var_type = _verify_and_get_type(cross_host_merge_result)
+                        else:
+                            if host_result:
+                                if cross_host_var_type == "dict":
+                                    cross_host_merge_result = self._merge_dict(host_result, cross_host_merge_result, [""])
+                                else:
+                                    cross_host_merge_result += host_result
+
+                ret.append(cross_host_merge_result)
```
Comment (m-a-r-k-e) [inline plugins/lookup/merge_variables.py:174]:
```suggestion
            else:  # consider variables of hosts in given groups
                cross_host_merge_result = initial_value
                for host in variables["hostvars"]:
                    if self._is_host_in_allowed_groups(variables["hostvars"][host]["group_names"]):
                        cross_host_merge_result = self._merge_vars(term, cross_host_merge_result, variables["hostvars"][host])

                ret.append(cross_host_merge_result)
```

By using the `initial_value` of the `_merge_vars` no duplication of the merge functionality is needed.
It looks like this evens solves a bug, see my comment for the integration test.

File: tests/integration/targets/lookup_merge_variables/test_all_hosts.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            msg: "{{ merged_dict }}"
+
+        - name: Test merge dicts via two of three groups - Validate that the dict is complete
+          ansible.builtin.assert:
+            that:
+              - "(merged_dict.keys() | list | length) == 3"
+              - "'item1' in merged_dict"
+              - "'item2' in merged_dict"
+              - "'list_item' in merged_dict"
+              - "merged_dict.list_item | length == 2"
```
Comment (m-a-r-k-e) [inline tests/integration/targets/lookup_merge_variables/test_all_hosts.yml:45]: An integration test that uses the inital value functionality is missing. I did a test using:

```yaml          
    - name: Test merge dicts via two of three groups with inital value
      delegate_to: localhost
      vars:
        initial_dict:
          initial: initial_value
        merged_dict: "{{ lookup(community.general.'merge_variables', '__merge_dict_in', initial_value=initial_dict, pattern_type='suffix', groups=['dummy1', 'dummy2']) }}"
      block:
        - name: Test merge dicts via two of three groups with inital value - Print the merged dict
          ansible.builtin.debug:
            msg: "{{ merged_dict }}"

        - name: Test merge dicts via two of three groups with inital value - Validate that the dict is complete
          ansible.builtin.assert:
            that:
              - "(merged_dict.keys() | list | length) == 4"
              - "'item1' in merged_dict"
              - "'item2' in merged_dict"
              - "'list_item' in merged_dict"
              - "merged_dict.list_item | length == 2"
```

This even shows an issue with the current implementation, the result was:

```json
{
    "initial": "initial_value",
    "item1": "value1",
    "item1": "value2",
    "list_item": [
        "test1",
        "test2"
        "test1"
        "test2"
    ]
}
```

But the expected result is:
```json
{
    "initial": "initial_value",
    "item1": "value1",
    "item1": "value2",
    "list_item": [
        "test1",
        "test2"
    ]
}
```

The suggestion to make the implementation easier, also solves this issue.

**Review by m-a-r-k-e — CHANGES_REQUESTED** [review]

File: tests/integration/targets/lookup_merge_variables/test_inventory_all_hosts.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+      testdict3__merge_dict_in:
+        item2: value2
+        list_item:
+          - test2
+    host3:
+      testdict3__merge_dict_in:
+        item3: value3
+        list_item:
+          - test3
```
[the remark this answers is not in the document]
Comment (alpex8) [reply tests/integration/targets/lookup_merge_variables/test_inventory_all_hosts.yml]: That's a good point! I changed the inventory accordingly

File: plugins/lookup/merge_variables.py (hunk 11 above)
Comment (alpex8) [reply plugins/lookup/merge_variables.py]: Thanks, that a nice improvement! I didn't like the duplication in the first place, but I didn't think of just reusing  intermediate is enough here.

File: tests/integration/targets/lookup_merge_variables/test_all_hosts.yml (hunk 12 above)
Comment (alpex8) [reply tests/integration/targets/lookup_merge_variables/test_all_hosts.yml:45]: Thanks, I took it over.

**Comment by rlenferink** [comment]
Looks good to me! Thanks for processing the review comments!

shipit

**Review by m-a-r-k-e — APPROVED** [review]
Thanks for processing the review comments, looks good now!

**Comment by felixfontein** [comment]
@alpex8 thanks for your contribution!
@rlenferink @m-a-r-k-e thanks for reviewing this!

**Pull request merged**